### PR TITLE
Gray Code of input length

### LINKED ISSUE #1644: Gray Code Generation of any input length
Would like to implement for GSSOC'20.Please assign me this issue.

### PR DESCRIPTION
**Pull request by SubhradeepSS**
Fixes #1644 

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- Added the code which prints the Gray Code sequence of any input length given as input by user.
- 

#### Languages Used:
- C++
- 

#### Files Added:
- [Gray Code.cpp](https://github.com/SubhradeepSS/Algo_Ds_Notes/blob/Gray-Code/Gray%20Code.cpp)
- 

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

File: Gray Code.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
+#include<bits/stdc++.h>
+using namespace std;
+
+void Gray_Code (int n)
+{
+  if (n <= 0)
```
Comment (jainaman224) [inline Gray Code.cpp]: You are not following 4 space indentation.

File: Gray Code.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    {
+      cout << "Invalid input length";
+      return;
+    }
+  vector < string > v;
```
Comment (jainaman224) [inline Gray Code.cpp]: Please indent code properly.

File: Gray Code.cpp (hunk 1 above)
Comment (SubhradeepSS) [reply Gray Code.cpp]: Spaces should be replaced by tab only right?

**Comment by jainaman224** [comment]
@SubhradeepSS Are you working on it?

**Comment by SubhradeepSS** [comment]
> @SubhradeepSS Are you working on it?

@jainaman224 Please clear the doubts I asked in the PR of Huffman Coding #1964

File: Gray Code.cpp (hunk 1 above)
Comment (jainaman224) [reply Gray Code.cpp]: 2 spaces should be replaced by 4 spaces.

**Comment by SubhradeepSS** [comment]
@jainaman224 Is the file properly formatted now?

**Comment by SubhradeepSS** [comment]
> 
> 
> @SubhradeepSS Are you working on it?

@jainaman224 Please review my PR

**Comment by SubhradeepSS** [comment]
@jainaman224 Please review my PR

**Comment by somya-kapoor** [comment]
@SubhradeepSS  Please use 4 space indentation. For reference visit any file present in the repository.

**Comment by SubhradeepSS** [comment]
> 
> 
> @SubhradeepSS Please use 4 space indentation. For reference visit any file present in the repository.

Is the file properly formatted now?

File: Gray Code.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    }
+    vector<string> v;
+    v.push_back("0");
+    v.push_back("1");
+    for(int i = 2; i < (1 << n); i <<= 1)
+    {
+        for(int j = i-1; j >= 0; j--)
+	    v.push_back(v[j]);
+        for(int j = 0; j < i;    j++)
+	    v[j] = "0" + v[j];
```
Comment (somya-kapoor) [inline Gray Code.cpp]: inconsistency in indentation

File: Gray Code.cpp [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        for(int    j = i;    j < 2 * i;    j++)
```
Comment (somya-kapoor) [inline Gray Code.cpp]: inconsistency in space

File: Gray Code.cpp [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+	    v[j] = "1" + v[j];
```
Comment (somya-kapoor) [inline Gray Code.cpp]: inconsistency in indentation

**Comment by SubhradeepSS** [comment]
@jainaman224 @somya-kapoor Please review

**Comment by SubhradeepSS** [comment]
@jainaman224 @somya-kapoor Please review

File: Gray Code.cpp [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        cout << "Invalid input length";
+        return;
+    }
+    vector<string> v;
+    v.push_back("0");
+    v.push_back("1");
+    for(int i = 2; i < (1 << n); i <<= 1)
+    {
+        for(int j = i-1; j >= 0; j--)
```
Comment (somya-kapoor) [inline Gray Code.cpp]: Space around the operator

File: Gray Code.cpp [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	    v.push_back(v[j]);
```
Comment (somya-kapoor) [inline Gray Code.cpp]: inconsistency in indentation

File: Gray Code.cpp (hunk 3 above)
Comment (SubhradeepSS) [reply Gray Code.cpp]: Can you please open the code file in my branch.I have properly indented it but it's not showing in the files changed section.Kindly see the actual file and then review

File: Gray Code.cpp (hunk 7 above)
Comment (SubhradeepSS) [reply Gray Code.cpp]: Can you please open the code file in my branch.I have properly indented it but it's not showing in the files changed section.Kindly see the actual file and then review.

Comment (SubhradeepSS) [reply Gray Code.cpp]: Capture

Comment (SubhradeepSS) [reply Gray Code.cpp]: > 
> 
> Capture

This is the screenshot of the part of the code you told to change.

Comment (SubhradeepSS) [reply Gray Code.cpp:16]: @somya-kapoor @jainaman224 Please review

**Comment by somya-kapoor** [comment]
@SubhradeepSS Change the indent type to space and indent size to 4 in Github file

**Comment by SubhradeepSS** [comment]
> 
> 
> @SubhradeepSS Change the indent type to space and indent size to 4 in Github file

Sorry,I didn't understand for what you said?

File: Gray Code.cpp (hunk 7 above)
Comment (somya-kapoor) [reply Gray Code.cpp:16]: Screenshot (108)

Comment (SubhradeepSS) [reply Gray Code.cpp]: @somya-kapoor Is it ok now?

File: Gray Code.cpp [hunk 8]
```diff
@@ -0,0 +1,53 @@
+#include<bits/stdc++.h>
```
Comment (somya-kapoor) [inline Gray Code.cpp:1]: Change the file name to Gray_Code/Gray_Code.cpp

**Comment by somya-kapoor** [comment]
Squash the commits after changing.

**Comment by somya-kapoor** [comment]
Include _ in the file name(Gray_Code/Gray_Code.cpp)

**Comment by SubhradeepSS** [comment]
> 
> 
> Squash the commits after changing.

@somya-kapoor I have squashed

File: Gray Code/Gray Code.cpp [hunk 9]
```diff
@@ -0,0 +1,53 @@
+#include<bits/stdc++.h>
```
Comment (somya-kapoor) [inline Gray Code/Gray Code.cpp:1]: File name should be  Gray_Code/Gray_Code.cpp

Comment (somya-kapoor) [reply Gray Code/Gray Code.cpp:1]: _ is included.

Comment (SubhradeepSS) [reply Gray Code/Gray Code.cpp]: done @somya-kapoor

**Comment by SubhradeepSS** [comment]
@somya-kapoor @jainaman224 Please review my PR.Squashing is done

**Comment by SubhradeepSS** [comment]
@somya-kapoor Please review my closed PR #1956 for the issue #1478 which was assigned to me to be implemented in c++.
Please look into this pr which was closed without notifying me.I made the changes.
Capture

**Comment by SubhradeepSS** [comment]
@somya-kapoor Is anything left to be done for the PR?

**Comment by SubhradeepSS** [comment]
@somya-kapoor While updating my branch with the original repo,I mistakenly deleted my earlier file.I updated it now.Will u check if everything is okay or not?

File: Gray_Code/Gray_Code.cpp [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    for(int i = 0; i < v.size(); i++)
+    	cout << v[i] << endl;
+}
+
+int main(){
+	int n;
+	cout << "Enter the bit length for Gray Code generation : ";
+	cin >> n;
+	Gray(n);
+}
```
Comment (somya-kapoor) [inline Gray_Code/Gray_Code.cpp:38]: Sample is missing

File: Gray_Code/Gray_Code.cpp [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    for(int i = 2; i < (1 << n); i <<= 1){
+    	for(int j = i - 1; j >= 0; j--){
+    	    v.pb(v[j]);
+    	}
+    	for(int j = 0; j < i; j++){
+    		v[j] = "0" + v[j];
+    	}
+    	for(int j = i; j < 2 * i; j++){
+    		v[j] = "1" + v[j];
+    	}
```
Comment (somya-kapoor) [inline Gray_Code/Gray_Code.cpp:24]: inconsistency in indentation

File: Gray_Code/Gray_Code.cpp [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+    for(int i = 2; i < (1 << n); i <<= 1){
+    	for(int j = i - 1; j >= 0; j--){
+    	    v.pb(v[j]);
+    	}
+    	for(int j = 0; j < i; j++){
+    		v[j] = "0" + v[j];
+    	}
+    	for(int j = i; j < 2 * i; j++){
+    		v[j] = "1" + v[j];
```
Comment (somya-kapoor) [inline Gray_Code/Gray_Code.cpp:23]: 4 space indentation

File: Gray_Code/Gray_Code.cpp [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    cout << "Gray Code for bit length = " << n << " : " << endl;
+
+    for(int i = 0; i < v.size(); i++)
+    	cout << v[i] << endl;
+}
+
+int main(){
+	int n;
```
Comment (somya-kapoor) [inline Gray_Code/Gray_Code.cpp:34]: 4 space indentation

File: Gray_Code/Gray_Code.cpp [hunk 14]
```diff
@@ -0,0 +1,38 @@
+#include<bits/stdc++.h>
```
Comment (somya-kapoor) [inline Gray_Code/Gray_Code.cpp:1]: Add comments

**Comment by SubhradeepSS** [comment]
@somya-kapoor Please review

**Comment by SubhradeepSS** [comment]
@somya-kapoor Please review my PR

**Comment by SubhradeepSS** [comment]
@somya-kapoor @jainaman224 Please review and tell what to update further?

**Comment by somya-kapoor** [comment]
Please squash the commits

**Comment by SubhradeepSS** [comment]
> 
> 
> Please squash the commits

@somya-kapoor Squashing is done.Please review now

**Comment by SubhradeepSS** [comment]
@somya-kapoor Please review this PR also

**Comment by SubhradeepSS** [comment]
@somya-kapoor review

File: Gray_Code/Gray_Code.cpp [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    vector<string> v;
+    v.pb("0");
+    v.pb("1");
+
+    for(int i = 2; i < (1 << n); i <<= 1){
+    	for(int j = i - 1; j >= 0; j--)
+    	    v.pb(v[j]);
+        
+    	for(int j = 0; j < i; j++)
+            v[j] = "0" + v[j];
```
Comment (singh-shreya6) [inline Gray_Code/Gray_Code.cpp:24]: Indentation is more than 4 spaces here.

File: Gray_Code/Gray_Code.cpp [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+        
+    	for(int j = i; j < 2 * i; j++)
+    	    v[j] = "1" + v[j];
+    }
+
+    cout << "Gray Code for bit length = " << n << " : " << endl;
+
+    for(int i = 0; i < v.size(); i++)
+    	cout << v[i] << endl;
```
Comment (singh-shreya6) [inline Gray_Code/Gray_Code.cpp:33]: Please add 4 space indentation here.

**Comment by somya-kapoor** [comment]
Are you working on it?

File: Gray_Code/Gray_Code.cpp (hunk 16 above)
Comment (SubhradeepSS) [reply Gray_Code/Gray_Code.cpp:33]: Capture
This is the screenshot of the file that I made.In the file I have done 4 space indentation but I dont understand why its not showing in the files changed seection.Also I have ensured the indentation in github as "Spaces" and "4".

Comment (SubhradeepSS) [reply Gray_Code/Gray_Code.cpp:33]: @singh-shreya6 I request you to kindly open the file in my branch and review it

**Comment by SubhradeepSS** [comment]
> 
> 
> Are you working on it?

Yes I am.But I dont understand why the problem of indentation is occuring.I request all of the reviewers:any one of them can kindly check the file in my branch where the 4 space indentation is clearly visible.

**Comment by SubhradeepSS** [comment]
> I looked into your branch. It looked correct, however when I copied your code to my editor and pasted back, indentation has become correct. Must be issue of spaces/tabs at some place or the other.
> I have made one commit to your branch.
> The code looks good now. Please squash your commits so that it can be merged.

@singh-shreya6 I have squashed the commits

**Comment by SubhradeepSS** [comment]
@singh-shreya6 Please have a look at my other PR #1964

**Pull request merged**